src/ctapipe/core/tool.py

```diff
@@ -429,46 +430,50 @@ def run(self, argv=None, raises=False):
+                current_exception = err
                 self.log.error("%s", err)
                 self.log.error("Use --help for more info")
                 exit_status = 2  # wrong cmd line parameter
                 Provenance().finish_activity(
                     activity_name=self.name, status="error", exit_code=exit_status
                 )
-                if raises:
-                    raise
             except KeyboardInterrupt:
                 self.log.warning("WAS INTERRUPTED BY CTRL-C")
                 exit_status = 130  # Script terminated by Control-C
                 Provenance().finish_activity(
                     activity_name=self.name, status="interrupted", exit_code=exit_status
                 )
             except Exception as err:
-                self.log.exception("Caught unexpected exception: %s", err)
-                exit_status = 1  # any other error
+                current_exception = err
+                exit_status = getattr(err, "exit_code", 1)
+                if exit_status == 1:
+                    self.log.exception("Caught unexpected exception: %s", err)
+                else:
+                    self.log.error("Caught exception: %s", err)
                 Provenance().finish_activity(
                     activity_name=self.name, status="error", exit_code=exit_status
                 )
-                if raises:
-                    raise
             except SystemExit as err:
                 exit_status = err.code
-                # Do nothing if SystemExit was called with the exit code 0 (e.g. with -h option)
-                if exit_status != 0:
-                    if raises:
-                        raise  # do not re-intercept in tests
-                    else:
-                        self.log.exception(
-                            "Caught SystemExit with exit code %s", exit_status
-                        )
-                        Provenance().finish_activity(
-                            activity_name=self.name,
-                            status="error",
-                            exit_code=exit_status,
-                        )
+                if exit_status == 0:
+                    # Finish normally
+                    Provenance().finish_activity(activity_name=self.name)
+                else:
+                    # Finish with error
+                    current_exception = err
+                    self.log.critical(
+                        "Caught SystemExit with exit code %s", exit_status
+                    )
+                    Provenance().finish_activity(
+                        activity_name=self.name,
+                        status="error",
+                        exit_code=exit_status,
+                    )
             finally:
                 if not {"-h", "--help", "--help-all"}.intersection(self.argv):
                     self.write_provenance()
+                if raises and current_exception:
+                    raise current_exception
```

Comment:
Doesn't this destroy the traceback? A simple `raise` in the except block keeps the traceback chain intact, I am not sure, but I think raising the specific exception here in the finally block will destroy the traceback.

Reply:
Ok, nevermind, seems the traceback stays intact

Reply:
It has been a long time since a ran into a python 3 vs python 2 thing... in python 2, what I said was true.

Reply:
Yeah, it keeps the traceback, and moreover, a simple `raise` in the `finally` block will result in an error, as there will be no active exception to raise, that's why you can either have an `if raises: raise` clause in each leg, or carry the exception like this.